securesystemslib/signer/_key.py

```diff
@@ -0,0 +1,202 @@
+
+        return (
+            self.keyid == other.keyid
+            and self.keytype == other.keytype
+            and self.scheme == other.scheme
+            and self.keyval == other.keyval
+            and self.unrecognized_fields == other.unrecognized_fields
+        )
+
+    @classmethod
+    @abstractmethod
+    def from_dict(cls, keyid: str, key_dict: Dict[str, Any]) -> "Key":
+        """Creates ``Key`` object from a serialization dict
+
+        Key implementations must override this factory constructor that is used
+        as a deserialization helper.
+
+        Users should call Key.from_dict(): it dispatches to the actual subclass
+        implementation based on supported keys in KEY_FOR_TYPE_AND_SCHEME.
+
+        Raises:
+            KeyError, TypeError: Invalid arguments.
+        """
+        keytype = key_dict.get("keytype")
+        scheme = key_dict.get("scheme")
+        if (keytype, scheme) not in KEY_FOR_TYPE_AND_SCHEME:
+            raise ValueError(f"Unsupported public key {keytype}/{scheme}")
+
+        # NOTE: Explicitly not checking the keytype and scheme types to allow
+        # intoto to use (None,None) to lookup GPGKey, see issue #450
+        key_impl = KEY_FOR_TYPE_AND_SCHEME[(keytype, scheme)]  # type: ignore
+        return key_impl.from_dict(keyid, key_dict)
+
+    @abstractmethod
+    def to_dict(self) -> Dict[str, Any]:
+        """Returns a serialization dict.
+
+        Key implementations must override this serialization helper.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def verify_signature(self, signature: Signature, data: bytes) -> None:
+        """Raises if verification of signature over data fails.
+
+        Args:
+            signature: Signature object.
+            data: Payload bytes.
+
+        Raises:
+            UnverifiedSignatureError: Failed to verify signature.
+            VerificationError: Signature verification process error. If you
+                are only interested in the verify result, just handle
+                UnverifiedSignatureError: it contains VerificationError as well
+        """
+        raise NotImplementedError
+
+
+class SSlibKey(Key):
+    """Key implementation for RSA, Ed25519, ECDSA and Sphincs keys"""
+
+    def to_securesystemslib_key(self) -> Dict[str, Any]:
+        """Internal helper, returns a classic securesystemslib keydict"""
+        return {
+            "keyid": self.keyid,
+            "keytype": self.keytype,
+            "scheme": self.scheme,
+            "keyval": self.keyval,
+        }
+
+    @classmethod
+    def from_securesystemslib_key(cls, key_dict: Dict[str, Any]) -> "SSlibKey":
+        """Constructor from classic securesystemslib keydict"""
+        # ensure possible private keys are not included in keyval
+        return SSlibKey(
+            key_dict["keyid"],
+            key_dict["keytype"],
+            key_dict["scheme"],
+            {"public": key_dict["keyval"]["public"]},
```

Comment:
this looks weird: I think I made a mistake...

Reply:
It's a *heisenpatch*